internal/xds/balancer/outlierdetection/balancer_test.go

```diff
@@ -1027,253 +1028,348 @@ func (s) TestDurationOfInterval(t *testing.T) {
+						}
+						scw.Connect()
+						allSubConns[i] = scw
+					}
 
-	// Register health listeners after all the connectivity updates are
-	// processed to avoid data races while accessing the health listener within
-	// the TestClientConn.
-	select {
-	case <-ctx.Done():
-		t.Fatal("Context timed out waiting for all SubConns to become READY.")
-	case <-connectivityCh:
-	}
+					bd.ClientConn.UpdateState(balancer.State{
+						ConnectivityState: connectivity.Ready,
+						Picker: &rrPicker{
+							scs: allSubConns,
+						},
+					})
+					return nil
+				},
+			})
 
-	scw1.RegisterHealthListener(func(healthState balancer.SubConnState) {
-		scsCh.Send(subConnWithState{sc: scw1, state: healthState})
-	})
-	scw2.RegisterHealthListener(func(healthState balancer.SubConnState) {
-		scsCh.Send(subConnWithState{sc: scw2, state: healthState})
-	})
-	scw3.RegisterHealthListener(func(healthState balancer.SubConnState) {
-		scsCh.Send(subConnWithState{sc: scw3, state: healthState})
-	})
+			od, tcc, cleanup := setup(t)
+			defer cleanup()
+			endpoints := make([]resolver.Endpoint, len(allSubConns))
+			for i := range allSubConns {
+				endpoints[i] = resolver.Endpoint{Addresses: []resolver.Address{{Addr: fmt.Sprintf("address%d", i+1)}}}
+			}
+			od.UpdateClientConnState(balancer.ClientConnState{
+				ResolverState: resolver.State{
+					Endpoints: endpoints,
+				},
+				BalancerConfig: &test.lbConfig,
+			})
 
-	select {
-	case <-ctx.Done():
-		t.Fatalf("timeout while waiting for a UpdateState call on the ClientConn")
-	case picker := <-tcc.NewPickerCh:
-		// Set each of the three upstream addresses to have five successes each.
-		// This should cause none of the addresses to be ejected as none of them
-		// are outliers according to the success rate algorithm.
-		for i := 0; i < 3; i++ {
-			pi, err := picker.Pick(balancer.PickInfo{})
-			if err != nil {
-				t.Fatalf("picker.Pick failed with error: %v", err)
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+			defer cancel()
+
+			// Transition the SubConns to READY so that they can register health
+			// listeners.
+			for range test.numberOfConns {
+				select {
+				case <-ctx.Done():
+					t.Fatalf("Timed out waiting for creation of new SubConn.")
+				case sc := <-tcc.NewSubConnCh:
+					sc.UpdateState(balancer.SubConnState{ConnectivityState: connectivity.Connecting})
+					sc.UpdateState(balancer.SubConnState{ConnectivityState: connectivity.Ready})
+				}
 			}
-			for c := 0; c < 5; c++ {
-				pi.Done(balancer.DoneInfo{})
+
+			// Register health listeners after all the connectivity updates are
+			// processed to avoid data races while accessing the health listener within
+			// the TestClientConn.
+			connectionsReady := 0
+		OuterLoop:
+			for {
+				select {
+				case <-ctx.Done():
+					t.Fatal("Context timed out waiting for all SubConns to become READY.")
+				case <-connectivityCh:
+					connectionsReady++
+					if connectionsReady == len(allSubConns) {
```

Comment:
If we change the `for` loop to be something like `for ; connectionsReady < test.numberOfConns; { ... }`, we wouldn't need the `OuterLoop` label right?

Reply:
yeah good idea, I was pretty tunnel focused on using `for select` that missed that fixed in 32d9eb3ea51d51cdcf5c8ae1477f6c435ed878bd